_posts/2020-05-25-kinesis-streams-part-1.md

```diff
@@ -0,0 +1,439 @@
+      }, delay(retryCount))
+    } else if (retryCount >= MAX_RETRIES) {
+      sendToDLQ(streamName, retryCount)
+    }
+  })
+}
+ 
+function delay(attempt) {
+  return RETRY_BASE_DELAY * (2 ** attempt)
+}
+```
+
+That is already pretty good, compared to where we started, right? However, we can do even better than that! To avoid sending bursts of retries, we should add a random *jitter* to the delay. In that way, we will spread the retry attempts more uniformly and significantly decrease the risk of overwhelming the stream. This is what our `delay` function might look like:
+
+```javascript
+function delay(attempt) {
+  const exponentialDelay = RETRY_BASE_DELAY * (2 ** attempt)
+  return Math.floor(Math.random() * (exponentialDelay - RETRY_BASE_DELAY)) + RETRY_BASE_DELAY
+}
+```
+
+And we are done! We have built a simple, yet effective retry strategy to handle partial failures.
+
+There is an [excellent article](https://aws.amazon.com/builders-library/timeouts-retries-and-backoff-with-jitter/?did=ba_card&trk=ba_card) about retries and timeouts in the Amazon Builders' Library that I would strongly encourage you to read. This is incredibly helpful because it is not specific to any particular service, but rather to the nature of distributed systems.
+
+And speaking of timeouts...
+
+
+## Timeouts
+
+Any remote call can timeout. This can happen due to a multitude of reasons, such as network glitches, service being busy, or the response being lost on its way to you in the vast space of the Internet. This happens all the time and this happens to AWS API calls as well. That's why AWS SDK has a **default value set for request timeouts**. After the timeout is reached, a request will fail and will be retried automatically, like any other retriable failed request.
+
+Each SDK has its own ways of configuring the request timeouts. Node.js SDK allows you to set two distinct types of timeous:
+- `connectTimeout`: a timeout for establishing a **new connection** on a socket; if not explicitly set, this value will default to the value of the `timeout`
+- `timeout`: read timeout for an existing socket, or, in other words, the time between when the request ends and the response is received, **including service and network round-trips**
+
+It's important to understand the differences between the two. It's equally important to know what the default values are and to configure the timeout values according to the expected latencies of the services you are using.
+
+In the case of Kinesis API, the default timeout on an existing socket (the `timeout` configuration) is **2 minutes**. This means that, by default, AWS SDK will wait for the response for the entire two minutes, before it gives up and retries the request. Add to that the default 2 minutes for establishing a new connection and possible retries that follow both of the timeouts, and you get a producer that is stuck for a while waiting for something that might never happen.
```

Comment:
timeout on an existing socket -> timeout of an existing socket

Reply:
I think for would be even better!